src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexCache.cs

```diff
@@ -0,0 +1,187 @@
+        /// Retrieves address data, either the cached version if it exists,
+        /// or directly from the underlying database.
+        /// </summary>
+        /// <param name="address">The address to retrieve data for.</param>
+        public AddressIndexerData GetAddress(string address)
+        {
+            lock (this.lockObj)
+            {
+                this.cachedAddresses.TryGetValue(address, out AddressIndexerData indexData);
+
+                if (indexData != null)
+                    return indexData;
+
+                indexData = this.addressIndexerData.FindById(address);
+
+                if (indexData == null)
+                    indexData = new AddressIndexerData() { Address = address, BalanceChanges = new List<AddressBalanceChange>() };
+
+                this.AddToCache(indexData);
+
+                return indexData;
+            }
+        }
+
+        /// <summary>
+        /// Checks for addresses that are affected by balance changes above a given block height.
+        /// This method should only be relied upon for block heights lower than the consensus tip and higher
+        /// than (tip - maxReorg). This is because it is only used while reorging the address indexer.
+        /// </summary>
+        /// <param name="height">The block height above which balance changes should be considered.</param>
+        /// <param name="finalisedHeight">Needed when flushing the cache.</param>
+        /// <returns>A list of affected addresses containing balance changes above the specified block height.</returns>
+        public List<string> GetAddressesHigherThanHeight(int height)
+        {
+            // Need to flush the cache first before searching the underlying database to ensure complete results.
+            this.Flush();
+
+            lock (this.lockObj)
+            {
+                // Need to specify index name explicitly so that it gets used for the query.
+                IEnumerable<AddressIndexerData> affectedAddresses = this.addressIndexerData.Find(Query.GT("BalanceChangedHeightIndex", height));
+
+                // Per LiteDb documentation:
+                // "Returning an IEnumerable your code still connected to datafile.
+                // Only when you finish consume all data, datafile will be disconnected."
+                return affectedAddresses.Select(x => x.Address).ToList();
+            }
+        }
+
+        public void Flush()
+        {
+            lock (this.lockObj)
+            {
+                var batch = new List<AddressIndexerData>();
+
+                foreach (string address in this.dirtyAddresses)
+                {
+                    // TODO: This seems like the only good place to collapse balance changes into fewer entries
+                    // TODO: The drawback with that being that unused indexer records on disk may then never get collapsed
+
+                    batch.Add(this.cachedAddresses[address]);
+
+                    if (batch.Count < BatchSize)
```

Comment:
Is 100 one of the "non-scientific" constants hehe ;) Is there any documentation on LiteDb and what it regards as good batch sizes for updating records?

Reply:
I'm not sure if there is anything specific, some reading of its closed issues indicates that it could possibly even be set in the region of several thousand. The optimal value would most likely depend on the underlying storage media, whether SSD or not.